### ISSUE TITLE
CNN training. Worse testing results with better training parameters 

### ISSUE THREAD

**Issue by don-prog** [davisking/dlib#1275]
I use dnn_metric_learning_on_images_ex.cpp for training face recognition neural net using dataset with ~300 000 persons and ~1 000 000 photos. 
When I used your default parameters for training except `samples_per_id` in `load_mini_batch()` which I set to 2 and `iterations_without_progress_threshold` which I set to 10 000, then I got ~0.08 loss, 280 num_right / 20 num_wrong results after training. Also I tested it in this way: I have a test dataset with ~14 000 persons with a some number of pictures each, so, I pick one photo of some person and sort all other photos of all persons in order of similarity using trained net. Net with 10 000 value of `iterations_without_progress_threshold` has a picked person on the 3(distance: 0.487641), 77(d.: 0.571242), 187(d.:0.600932) places(num of places - 200).

After that I trained another net with the same parameters except `iterations_without_progress_threshold` which I set to 40 000. Now I got ~0.015 loss, and worse results at the after training testing stage: ~227 num_right / ~73 num_wrong, but I also got better results with the nearest neighbours search test: 2(d.:0.214533), 3(d.:0.232513), 4(d.: 0.244046), 5(d.:0.278714), 11(d.:0.315623), 49(d.:0.356549) places(num of places - 200).

I think that I got worse results in the after training test, but better in the nearest neighbors search test due to the distance between all photos has decreased and now many photos are included in the threshold, though now the same person photos are closer to each other.

Should I correct something in the training process(and what should I change if yes) or set the suitable threshold at the testing stage would be the best solution?

**Comment by davisking** [comment]
I have no idea.  Try things and see what works.  I would try the settings I documented in the code, since those gave good results for the public dlib model.  Also make sure your training data has good labels. Images with incorrect labels will mess up the model.

**Comment by don-prog** [comment]
1. Do you mean all the dnn_trainer settings or those you used in the described example?
2. Is it normal and permissible situation when the threshold is set after training and doesn't correspond to the training threshold? Or it is mean that I need to re-train the model(please note that it gives better results, but with the shifted threshold)?

**Comment by davisking** [comment]
Yes, read the comments in this example:
https://github.com/davisking/dlib/blob/master/examples/dnn_face_recognition_ex.cpp.
It is all explained there.
